[AArch64] Teach areMemAccessesTriviallyDisjoint about scalable widths.

davemgreen opened the pull request:
The base change here is to change getMemOperandWithOffsetWidth to return a TypeSize Width, which in turn allows areMemAccessesTriviallyDisjoint to reason about trivially disjoint widths.

harviniriawan commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -3918,21 +3924,21 @@ bool AArch64InstrInfo::getMemOpInfo(unsigned Opcode, TypeSize &Scale,
   case AArch64::LD1RW_IMM:
   case AArch64::LD1RW_D_IMM:
   case AArch64::LD1RSW_IMM:
     Scale = TypeSize::getFixed(4);
```
aren't LD1RSW* SVE instructions as well? Any reasons why we don't change them to getScalable?

davemgreen replied:
I believe they load a scalar and broadcast them to all lanes, so only load a fixed number of bytes is read from memory.

harviniriawan replied:
ah ok yeah makes  sense

harviniriawan commented:
lgtm

david-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -1214,8 +1214,7 @@ static MachineBasicBlock::iterator convertCalleeSaveRestoreToSPPrePostIncDec(
 
-  TypeSize Scale = TypeSize::getFixed(1);
-  unsigned Width;
+  TypeSize Scale = TypeSize::getFixed(1), Width(0, false);
```
I think it would be nice to use consistent notation here and have 

 TypeSize Scale = TypeSize::getFixed(1), Width = TypeSize::getFixed(0);

david-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -2675,9 +2676,15 @@ bool AArch64InstrInfo::getMemOperandsWithOffsetWidth(
-                                    Width, TRI))
+                                    WidthN, TRI))
     return false;
+  // The maximum vscale is 16 undef AArch64, return the maximal extent for the
```
Should this comment be

  // The maximum vscale is 16 for AArch64

?

david-arm commented on llvm/test/CodeGen/AArch64/sve-aliasing.ll:
Do we have tests for loads and stores of predicate vectors, i.e. <vscale x 16 x i1>, <vscale x 8 x i1>, <vscale x 4 x i1>, <vscale x 2 x i1>? I'd expect aliasing information to be useful for those cases too.

davemgreen replied:
We don't seem to be able to generate offset versions of these instructions at the moment (except for stack spills/reloads), so I've made it a mir test.

david-arm reviewed (commented):
This looks like a nice improvement to alias analysis! I just had a few minor comments ...

davemgreen committed:
[AArch64] Teach areMemAccessesTriviallyDisjoint about scalable widths.

The base change here is to change getMemOperandWithOffsetWidth to return a
TypeSize Width, which in turn allows areMemAccessesTriviallyDisjoint to reason
about trivially disjoint widths.

david-arm reviewed (approved):
LGTM! Thanks for making the changes.
